docs/source/tutorials/websockets.rst

```diff
@@ -0,0 +1,1074 @@
+
+   app = Chalice(app_name="echo-server")
+   app.websocket_api.session = Session()
+   app.experimental_feature_flags.update([
+       'WEBSOCKETS'
+   ])
+
+
+   @app.on_ws_message()
+   def message(event):
+   try:
+       app.websocket_api.send(
+           connection_id=event.connection_id,
+           message=event.body,
+       )
+   except WebsocketDisconnectedError as e:
+       pass  # Disconnected so we can't send the message back.
+
+
+Explanation
+~~~~~~~~~~~
+
+Stepping through this app line by line, the first thing to note is that we
+need to import and instantiate a boto3 session. This session is manually
+assigned to ``app.websocket_api.session``.
+This is needed because in order to send websocket responses to API Gateway we
+need to construct a boto3 client. Chalice does not take a direct dependency
+on boto3 or botocore, so we need to provide the Session ourselves.
+
+.. code-block:: python
+
+   from boto3.session import Session
+   app.websocket_api.session = Session()
+
+
+Next we enable the experimental feature ``WEBSOCKETS``. As noted at the top
+of this section, websockets are an experimental feature and are subject to
+API changes.
+
+.. code-block:: python
+
+   app.experimental_feature_flags.update([
+       'WEBSOCKETS'
+   ])
+
+
+To register a websocket handler, and cause Chalice to deploy an
+API Gateway Websocket API we use the ``app.on_ws_message()`` decorator.
+The event parameter here is a wrapper object with some convenience
+parameters attached. The most useful are ``event.connection_id`` and
+``event.body``. The ``connection_id`` is an API Gateway specific identifier
+that allows you to refer to the connection that sent the message. The ``body``
+is the content of the message.
+
+.. code-block:: python
+
+   @app.on_ws_message()
+   def message(event):
+
+
+Since this is an echo server, the message handler simply reads the content it
+received on the socket, and rewrites it back to the same socket. To send a
+message to a socket we call ``app.websocket_api.send(connection_id, message)``.
+In this case, we just use the same ``connection_id`` we got the message from,
+and use the ``body`` we got from the event as the ``message`` to send.
+
+.. code-block:: python
+
+   app.websocket_api.send(
+       connection_id=event.connection_id,
+       message=event.body,
+    )
+
+
+Finally, we catch the exception ``WebsocketDisconnectedError`` which is raised
+by ``app.websocket_api.send`` if the provided ``connection_id`` is not
+connected anymore. In our case this doesn't really matter since we don't have
+anything tracking our connections. The error has a ``connection_id`` property
+that contains the offending connection id.
+
+.. code-block:: python
+
+   except WebsocketDisconnectedError as e:
+       pass  # Disconnected so we can't send the message back.
+
+
+To test out the echo server you can install ``websocket-client`` from PyPI::
+
+  pip install websocket-client
```

Comment:
I was getting SSL cert errors on python3.7, but python2.7 worked fine.  Even tried updating their cacert.pem file they bundle but still didn't work for python 3.7.  It'd be good to see if you're seeing this error and if so, see if we need use something else.

Reply:
I had no problems on 3.7.